### PR TITLE
Fix `jsonData` not passed correctly

### PR DESCRIPTION
**Pull request by JR-1991**
**Overview**

This PR addresses the issue reported at https://github.com/datalad/datalad-dataverse/issues/320, where metadata sent via NativeApi.replace_datafile wasn’t updating the file metadata correctly. The problem arose because the jsonData payload needs to be sent as form-data, which HTTPX’s json kwarg doesn’t handle properly.

The fix involves sending the payload via the data argument. However, since other endpoints require data to be sent through the json argument, this PR introduces a private method in api.py to determine the correct way to send the data. This method checks for the presence of the jsonData key and adjusts the argument accordingly. If the key isn’t found, the json argument is used by default.

To prevent similar issues in the future, the tests have been updated. The replace_datafile tests now ensure that the metadata is correctly updated.

**TLDR**

* If `jsonData` key is present sent through `data` kwarg
* If not, JSON payload is sent through `json`
* Extended tests to verify metadata is updated correctly

### CODE REVIEW

**Comment by mih** [comment]
I can confirm that this change makes the previously failing test in `datalad-dataverse` pass again. Thank you!

**Comment by pdurbin** [comment]
Nice work, @JR-1991!

**Comment by shoeffner** [comment]
Please note @adswa's changes in  datalad/datalad-dataverse#322 and my alternative fix in #207, which is a PR onto this branch, not the main branch.

**Comment by JR-1991** [comment]
@shoeffner, thanks for the notice and the PR! Once I return from vacation, I will add a test to check for empty `json_str` to this PR.

**Comment by JR-1991** [comment]
I've included a test case to verify whether the behavior is as expected when `json_str` is set to `None` in the case of uploading a file. Would either @shoeffner or @pdurbin be available to review the PR?

**Comment by pdurbin** [comment]
@shoeffner are you interested in "triage" access? I just sent you an invite. You should see it at https://github.com/gdcc 😄

File: pyDataverse/api.py [hunk 1]
```diff
@@ -216,19 +215,22 @@ def put_request(self, url, data=None, auth=False, params=None):
         if isinstance(data, str):
             data = json.loads(data)
 
+        # Decide whether to use 'data' or 'json' args
+        request_params = self._check_json_data_form(data)
```
Comment (shoeffner) [inline pyDataverse/api.py:219]: I was thinking about calling it with `Api._check_json_data_form` since it's declared `@staticmethod`, but `self` works of course as well and allows for better inheritance.
Maybe it makes sense to move it outside of the Api class? Makes it also much easier to (unit-)test it in isolation.

But I guess this is more or less a matter of personal preference and thus I'd keep it as is. This way it's encapsulated within the API and it would be awkward to call it from other contexts, which I think is good.

**Review by shoeffner — APPROVED** [review]
I haven't run the tests, so I trust you that it works :)

This looks good and I just added a few comments as food for thought here and there. In general, I would merge it and then see what we can do in #189 to improve further.
Right now, this looks like a very good fix for the problem at hand.
Thanks!

:shipit:

File: tests/api/test_upload.py [hunk 2]
```diff
@@ -141,6 +176,7 @@ def test_file_replacement(self):
                 "description": "My description.",
                 "categories": ["Data"],
                 "forceReplace": False,
+                "directoryLabel": "some/other",
```
[the remark this answers is not in the document]
Comment (JR-1991) [reply tests/api/test_upload.py:179]: Makes sense! So you would suggest splitting this test into two separate, where one is testing the sole replacement of a file and the other in combination with changes in metadata?

File: tests/api/test_upload.py [hunk 3]
```diff
@@ -45,6 +45,41 @@ def test_file_upload(self):
...
+
+        # Act
+        response = api.upload_datafile(
+            identifier=pid,
+            filename="tests/data/datafile.txt",
+            json_str=None,
+        )
+
+        # Assert
+        assert response.status_code == 200, "File upload failed."
```
[the remark this answers is not in the document]
Comment (JR-1991) [reply tests/api/test_upload.py:81]: I agree there should be a check beyond the status code. I think it is better to take care of it now than later. Will extend the test(s) to check for content and correct metadata.

File: tests/api/test_upload.py [hunk 4]
```diff
@@ -246,3 +291,29 @@ def _fetch_datafile_content(
...
+
+        Args:
+            BASE_URL (str): The base URL of the Dataverse instance.
+            API_TOKEN (str): The API token for authentication.
+            id (str): The ID of the file.
+
+        Returns:
+            dict: The metadata for the file.
+        """
+        response = httpx.get(
```
[the remark this answers is not in the document]
Comment (JR-1991) [reply tests/api/test_upload.py:312]: I wanted to make sure to isolate the method as good as possible and not depend on other parts of the library. But maybe I am overthinking this. Would you suggest simply using the API method then?

File: tests/api/test_upload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shoeffner) [reply tests/api/test_upload.py:179]: Best would be all three, I guess: replacement of file, replacement of metadata, replacement of both.
I am not sure if "replacement of metadata" in isolation works though, see https://github.com/datalad/datalad-dataverse/issues/320#issuecomment-2248988896 where I analyzed that replacing a file with itself is not possible.

File: tests/api/test_upload.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (shoeffner) [reply tests/api/test_upload.py:312]: Yes, this is otherwise a place we might miss when things must be updated. But I am fine with having the call like this.

File: tests/api/test_upload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JR-1991) [reply tests/api/test_upload.py:179]: To my knowledge, only replacing the metadata is not possible since you have to supply a file with new content, as you already highlighted. One option would be to use the `forceReplace` prop. This should override the content check, but I am not 100% sure.

File: tests/api/test_upload.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (JR-1991) [reply tests/api/test_upload.py:312]: Alright, I will then use the API endpoint of pyDataverse instead.

File: pyDataverse/api.py [hunk 5]
```diff
@@ -268,6 +270,33 @@ def delete_request(self, url, auth=False, params=None):
                 params=params,
             )
 
+    @staticmethod
+    def _check_json_data_form(data: Optional[Dict]):
```
[the remark this answers is not in the document]
Comment (JR-1991) [reply pyDataverse/api.py:274]: Good point. Shall we put that into an issue for the future and further discuss it?

File: tests/api/test_upload.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (shoeffner) [reply tests/api/test_upload.py:179]: Then I guess having both – replacing the file + replacing the file and metadata – is the way to go.

File: pyDataverse/api.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (shoeffner) [reply pyDataverse/api.py:274]: I don't think there's a need for that; let's first see what comes out of #189 :)

**Comment by JR-1991** [comment]
@shoeffner I have updated this PR and specifically the tests to use pyDataverse methods instead of manual requests, as highlighted by you. The PR is now ready for review.

**Review by pdurbin — APPROVED** [review]
I didn't test this but tests are passing (and there are lots of new tests!) and the code makes sense.

**Comment by shoeffner** [comment]
Sorry for not responding earlier, but it looks good and I am happy it got merged!
